### PR TITLE
Fix bug in cubic B spline and enhance class

### PR DESCRIPTION
**Pull request by florent-lamiraux**
Add a constructor with a vector of knots,
fix bug in computation of basis functions. The bug has no effect when knots are equally spaced.

### CODE REVIEW

**Comment by bchretien** [comment]
:+1: Looks good, thanks!

However a test is missing for the new constructor. An example with non-uniform b-splines would probably be the best thing to do, since this was not really tested (and fixed) until now. Copying and adapting the `b-spline` and/or `cubic-b-spline` tests for the non-uniform case should be a good start.

**Comment by florent-lamiraux** [comment]
I have added a test that generates basis functions of non-uniform cubic B splines.
The output of the test is understood by gnuplot. The curves look correct.
There are floating point comparisons between the values returned by the basis functions and values stored in a local file.

File: src/cubic-b-spline.cc [hunk 1]
```diff
@@ -162,41 +182,26 @@ namespace trajectory
...
     std::size_t i_ = static_cast<std::size_t> (i);
-    size_type iPrev = 0;
 
-    while (!found && iPrev != i)
+    while (!found)
       {
 	i = Double2SizeType::convert
-	  (std::floor (static_cast<double> (imin) + (t - tmin)
-		       / (tmax - tmin) * static_cast<double> (imax - imin)));
+	  (std::floor (static_cast<double> (.5*(imin + imax)+.5)));
```
Comment (bchretien) [inline src/cubic-b-spline.cc]: > /home/travis/build/roboptim/roboptim-trajectory/src/cubic-b-spline.cc:196:53: error: conversion to ‘double’ from ‘roboptim::trajectory::Trajectory<3u>::size_type {aka long int}’ may alter its value [-Werror=conversion]

This should probably be:

```cpp
(std::floor (.5 * static_cast<double> (imin + imax) + .5));
```

**Comment by florent-lamiraux** [comment]
I fixed the warning. It compiles on my machine.
Let see what travis thinks about it.

**Comment by bchretien** [comment]
Looks good. Just a question regarding d6c48e1 (first commit for the gradient test): was it related to a difference in the output for a different version of Eigen? Or a lack of precision in the output?

**Comment by florent-lamiraux** [comment]
The test failed because of numerical precision. At several places, the last decimal was different. I do not know precisely the reason, eigen version, different os...

**Comment by bchretien** [comment]
I see, thanks for testing and fixing it then, these errors can be difficult to predict.

**Pull request merged**